added service catalog cli details

gaurav-nelson opened the pull request:
For https://trello.com/c/0EVVA4po/855-document-catalog-310-add-svcat-cli-to-service-catalog-in-openshift

Preview: https://service-catalog-cli--gnelson-preview.netlify.com/openshift-enterprise/(head%20detached%20at%20fetch_head)/architecture/service_catalog/service_catalog_cli.html

gaurav-nelson commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Document: [catalog-3.10] Add svcat CLI to service-catalog in OpenShift](https://trello.com/c/0EVVA4po/855-document-catalog-310-add-svcat-cli-to-service-catalog-in-openshift)

gaurav-nelson commented:
@jpeeler PTAL

gaurav-nelson commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,434 @@
+
+[[service-catalog-cli-broker-details]]
+=== Get broker details
+TBD
```
To be done...

sunzhaohua2 commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,434 @@
+[[service-catalog-cli-usage]]
+== Using `svcat`
+This section includes common commands to handle the user associated tasks listed
+in xref:index.adoc#service-catalog-design[the service catalog workflow]. Use the `svcat help` command to get more information and view other available
```
"svcat help"  change to "svcat --help"

sunzhaohua2 commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,434 @@
+----
+<1> Replace `<project-name>` with name of your project.
+
+. Create a service instance definition file *_postgresql-apb.yaml_*.
```
I think we should better use svcat command. For example, 
```
$ svcat provision -n szh-project postgresql-instance --class rh-postgresql-apb --plan dev --params-json '{"postgresql_database":"admin","postgresql_password":"admin","postgresql_user":"admin","postgresql_version":"9.6"}'
  Name:        postgresql-instance  
  Namespace:   szh-project          
  Status:                           
  Class:       rh-postgresql-apb    
  Plan:        dev                  

Parameters:
  postgresql_database: admin
  postgresql_password: admin
  postgresql_user: admin
  postgresql_version: "9.6"
```

sunzhaohua2 commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,434 @@
+. The service catalog controller adds those credentials as secrets to the project.
+====
+
+. Create a service binding definition file *_postgresql-apb-mediawiki-apb-bind.yaml_*
```
Also I think we should better use svcat command. For example, 
```
$ svcat bind postgresql-instance --name mediawiki-postgresql-binding
  Name:        mediawiki-postgresql-binding  
  Namespace:   szh-project                   
  Status:                                    
  Secret:      mediawiki-postgresql-binding  
  Instance:    postgresql-instance           

Parameters:
  {}

$ svcat get bindings
              NAME                NAMESPACE         INSTANCE         STATUS  
+------------------------------+-------------+---------------------+--------+
  mediawiki-postgresql-binding   szh-project   postgresql-instance   Ready  
```

sunzhaohua2 commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,434 @@
++
+[source, bash]
+----
+$ oc delete -n <project-name> servicebindings <binding-name>
```
Unbind the service instance is ok, we don't need to delete the service binding in step 2.  
```
$ svcat unbind --name mediawiki-postgresql-binding
deleted mediawiki-postgresql-binding
```

sunzhaohua2 commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,434 @@
+$ svcat unbind -n <project-name> <instance-name>
+----
+
+. Delete the service instance:
```
Deprovision the service instance is ok, we don't need to delete the service instance in step 2.  
```
$ svcat deprovision postgresql-instance
deleted postgresql-instance
```

sunzhaohua2 commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,434 @@
+$ oc delete clusterservicebrokers <broker-name>
+----
+
+To verify that the broker is removed, view the `ClusterServiceClass` resources for the broker.
```
I think we should verify the broker is removed first.

```
$ svcat get brokers
           NAME                                                URL                                        STATUS  
+------------------------+------------------------------------------------------------------------------+--------+
```

gaurav-nelson commented:
@jpeeler @sunzhaohua2 I was unable to test or run the commands myself. Can you please help me by providing example commands and output for following commands so that all the commands follow a single example through out.

1. `svcat get brokers`
2. `svcat sync broker <broker-name>`
3. `svcat get classes`
4. `svcat describe class <class-name>`
5. `svcat get plans`
6. `svcat get instances`

Please see the [rendered document](https://service-catalog-cli--gnelson-preview.netlify.com/openshift-enterprise/(head%20detached%20at%20fetch_head)/architecture/service_catalog/service_catalog_cli.html) for this PR to identify the commands and the flow.

sunzhaohua2 commented:
@gaurav-nelson Below are some examples you can refer. And delete bindings and instances using commands "svcat unbind" and "svcat deprovision"  are ok, we should remove "oc delete" command related. Thanks.
```
$ svcat get brokers
           NAME                                                        URL                                              STATUS  
+-------------------------+-------------------------------------------------------------------------------------------+--------+
  ansible-service-broker    https://asb.openshift-ansible-service-broker.svc:1338/ansible-service-broker                Ready   
  template-service-broker   https://apiserver.openshift-template-service-broker.svc:443/brokers/template.openshift.io   Ready   

$ svcat sync broker ansible-service-broker
Successfully fetched catalog entries from the ansible-service-broker broker

$ svcat get classes
                      NAME                                                                         DESCRIPTION                                                                     UUID                  
+-----------------------------------------------+----------------------------------------------------------------------------------------------------------------+--------------------------------------+
  rh-mediawiki-apb                                Mediawiki apb implementation                                                                                     03b69500305d9859bb9440d9f9023784      
  
...                                                          

  rh-mariadb-apb                                  Mariadb apb implementation                                                                                       2c259ddd8059b9bc65081e07bf20058f      
  rh-mysql-apb                                    Software Collections MySQL APB                                                                                   73ead67495322cc462794387fa9884f5      
  rh-postgresql-apb                               SCL PostgreSQL apb implementation                                                                                d5915e05b253df421efe6e41fb6a66ba      

$ svcat describe class rh-postgresql-apb
  Name:          rh-postgresql-apb                  
  Description:   SCL PostgreSQL apb implementation  
  UUID:          d5915e05b253df421efe6e41fb6a66ba   
  Status:        Active                             
  Tags:          database, postgresql               
  Broker:        ansible-service-broker             

Plans:
  NAME            DESCRIPTION            
+------+--------------------------------+
  prod   A single DB server with         
         persistent storage              
  dev    A single DB server with no      
         storage                         

$ svcat get plans
   NAME                         CLASS                                DESCRIPTION                             UUID                  
+---------+-----------------------------------------------+--------------------------------+--------------------------------------+
  default   rh-mediawiki-apb                                An APB that deploys MediaWiki    43d3e23d214c26dbebc0879e44425db4      
  default   mariadb-ephemeral                               Default plan                     09c9e2fb-6881-11e8-bac8-fa163e5ec98b 
   
...

  prod      rh-mariadb-apb                                  This plan deploys a single       93a78b607a2e18776587f0a35ebe92a6      
                                                            MariaDB instance with 10 GiB                                           
                                                            of persistent storage                                                  
  dev       rh-mariadb-apb                                  This plan deploys a single       ba144caf21607ad5d50c06bdd0f72db3      
                                                            MariaDB instance with                                                  
                                                            ephemeral storage                                                      
  prod      rh-mysql-apb                                    A MySQL server with persistent   465063705516630c3d3e2de7798b7041      
                                                            storage                                                                
  dev       rh-mysql-apb                                    A MySQL server with ephemeral    ac68a02cfc1148713f61edb99213a17d      
                                                            storage                                                                
  prod      rh-postgresql-apb                               A single DB server with          4acaf1511a92890cd8910b1d8473be97      
                                                            persistent storage                                                     
  dev       rh-postgresql-apb                               A single DB server with no       9783fc2e859f9179833a7dd003baa841      
                                                            storage                                                                

$ svcat describe plan rh-postgresql-apb/dev
  Name:          dev                                 
  Description:   A single DB server with no storage  
  UUID:          9783fc2e859f9179833a7dd003baa841    
  Status:        Active                              
  Free:          true                                
  Class:         rh-postgresql-apb                   

Instances:
No instances defined

Instance Create Parameter Schema:
  $schema: http://json-schema.org/draft-04/schema
  additionalProperties: false
  properties:
    postgresql_database:
      default: admin
      pattern: ^[a-zA-Z_][a-zA-Z0-9_]*$
      title: PostgreSQL Database Name
      type: string
    postgresql_password:
      pattern: ^[a-zA-Z0-9_~!@#$%^&*()-=<>,.?;:|]+$
      title: PostgreSQL Password
      type: string
    postgresql_user:
      default: admin
      maxLength: 63
      pattern: ^[a-zA-Z_][a-zA-Z0-9_]*$
      title: PostgreSQL User
      type: string
    postgresql_version:
      default: "9.6"
      enum:
      - "9.6"
      - "9.5"
      - "9.4"
      title: PostgreSQL Version
      type: string
  required:
  - postgresql_database
  - postgresql_user
  - postgresql_password
  - postgresql_version
  type: object

Instance Update Parameter Schema:
  $schema: http://json-schema.org/draft-04/schema
  additionalProperties: false
  properties:
    postgresql_version:
      default: "9.6"
      enum:
      - "9.6"
      - "9.5"
      - "9.4"
      title: PostgreSQL Version
      type: string
  required:
  - postgresql_version
  type: object

Binding Create Parameter Schema:
  $schema: http://json-schema.org/draft-04/schema
  additionalProperties: false
  type: object



$ svcat provision postgresql-instance --class rh-postgresql-apb --plan dev --params-json  '{"postgresql_database":"admin","postgresql_password":"admin","postgresql_user":"admin","postgresql_version":"9.6"}' -n szh-project
  Name:        postgresql-instance  
  Namespace:   szh-project          
  Status:                           
  Class:       rh-postgresql-apb    
  Plan:        dev                  

Parameters:
  postgresql_database: admin
  postgresql_password: admin
  postgresql_user: admin
  postgresql_version: "9.6"

$ svcat get instance
         NAME            NAMESPACE          CLASS         PLAN   STATUS  
+---------------------+-------------+-------------------+------+--------+
  postgresql-instance   szh-project   rh-postgresql-apb   dev    Ready   
  
  
$ svcat bind postgresql-instance --name mediawiki-postgresql-binding
  Name:        mediawiki-postgresql-binding  
  Namespace:   szh-project                   
  Status:                                    
  Instance:    postgresql-instance           

Parameters:
  {}

$ svcat get bindings
              NAME                NAMESPACE         INSTANCE         STATUS  
+------------------------------+-------------+---------------------+--------+
  mediawiki-postgresql-binding   szh-project   postgresql-instance   Ready   

$ svcat describe instance postgresql-instance
  Name:        postgresql-instance                                                                
  Namespace:   szh-project                                                                        
  Status:      Ready - The instance was provisioned successfully @ 2018-06-05 08:42:55 +0000 UTC  
  Class:       rh-postgresql-apb                                                                  
  Plan:        dev                                                                                

Parameters:
  postgresql_database: admin
  postgresql_password: admin
  postgresql_user: admin
  postgresql_version: "9.6"

Bindings:
              NAME               STATUS  
+------------------------------+--------+
  mediawiki-postgresql-binding   Ready   

$ oc get secret
NAME                           TYPE                                  DATA      AGE
builder-dockercfg-jxk48        kubernetes.io/dockercfg               1         6m
builder-token-92jrf            kubernetes.io/service-account-token   4         6m
builder-token-b4sm6            kubernetes.io/service-account-token   4         6m
default-dockercfg-cggcr        kubernetes.io/dockercfg               1         6m
default-token-g4sg7            kubernetes.io/service-account-token   4         6m
default-token-hvdpq            kubernetes.io/service-account-token   4         6m
deployer-dockercfg-wm8th       kubernetes.io/dockercfg               1         6m
deployer-token-hnk5w           kubernetes.io/service-account-token   4         6m
deployer-token-xfr7c           kubernetes.io/service-account-token   4         6m
mediawiki-postgresql-binding   Opaque                                6         3m

$ svcat unbind postgresql-instance -n szh-project
deleted mediawiki-postgresql-binding

$ oc get secret -n szh-project
NAME                       TYPE                                  DATA      AGE
builder-dockercfg-jxk48    kubernetes.io/dockercfg               1         9m
builder-token-92jrf        kubernetes.io/service-account-token   4         9m
builder-token-b4sm6        kubernetes.io/service-account-token   4         9m
default-dockercfg-cggcr    kubernetes.io/dockercfg               1         9m
default-token-g4sg7        kubernetes.io/service-account-token   4         9m
default-token-hvdpq        kubernetes.io/service-account-token   4         9m
deployer-dockercfg-wm8th   kubernetes.io/dockercfg               1         9m
deployer-token-hnk5w       kubernetes.io/service-account-token   4         9m
deployer-token-xfr7c       kubernetes.io/service-account-token   4         9m

$ svcat get bindings
  NAME   NAMESPACE   INSTANCE   STATUS  
+------+-----------+----------+--------+


$ svcat deprovision postgresql-instance
deleted postgresql-instance


$ svcat get instance
  NAME   NAMESPACE   CLASS   PLAN   STATUS  
+------+-----------+-------+------+--------+

$ oc delete clusterservicebrokers template-service-broker
clusterservicebroker "template-service-broker" deleted

$ svcat get brokers
           NAME                                                        URL                                              STATUS  
+-------------------------+-------------------------------------------------------------------------------------------+--------+
  ansible-service-broker    https://asb.openshift-ansible-service-broker.svc:1338/ansible-service-broker                Ready   

```

gaurav-nelson commented:
Thank you @sunzhaohua2 I will update this soon.

gaurav-nelson commented:
@sunzhaohua2 @jpeeler I have updated this doc, Can you please review this?

sunzhaohua2 commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,439 @@
+To view the available ClusterServiceClass resources:
+[source, bash]
+----
+$ svcat get classes
```
Sorry. This part need to update. The new version svcat doesn't include UUID in get class|plan output.
```
$ svcat get classes
        NAME                   DESCRIPTION            
+-------------------+--------------------------------+
  rh-mediawiki-apb    Mediawiki apb implementation    
  
  ...
  
  rh-mariadb-apb      Mariadb apb implementation      
  rh-mysql-apb        Software Collections MySQL APB  
  rh-postgresql-apb   SCL PostgreSQL apb              
                      implementation  
```

sunzhaohua2 commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,439 @@
+[source, bash]
+----
+$ svcat get plans
+   NAME                         CLASS                                DESCRIPTION                             UUID
```
This part also need to update. The new version svcat doesn't include UUID in get class|plan output.
```
$ svcat get plans
   NAME           CLASS                  DESCRIPTION            
+---------+-------------------+--------------------------------+
  default   rh-mediawiki-apb    An APB that deploys MediaWiki   
  
  ...
  
  prod      rh-mariadb-apb      This plan deploys a single      
                                MariaDB instance with 10 GiB    
                                of persistent storage           
  dev       rh-mariadb-apb      This plan deploys a single      
                                MariaDB instance with           
                                ephemeral storage               
  prod      rh-mysql-apb        A MySQL server with persistent  
                                storage                         
  dev       rh-mysql-apb        A MySQL server with ephemeral   
                                storage                         
  prod      rh-postgresql-apb   A single DB server with         
                                persistent storage              
  dev       rh-postgresql-apb   A single DB server with no      
                                storage   
```

sunzhaohua2 commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,439 @@
+To view service instance details, run the command:
+[source, bash]
+----
+svcat get instance
```
need to add "$ "

gaurav-nelson commented:
Thanks @sunzhaohua2 I have made suggested updates.

gaurav-nelson commented:
@openshift/team-documentation PTAL

bfallonf commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,438 @@
+The Service Catalog command-line interface (CLI) utility called *`svcat`* is
+available to handle these user related tasks. Although these tasks can be
+handled by various `oc` commands, `svcat`  provides simpler commands to do the
+same tasks, allowing easier interaction with the Service Catalog resources.
```
To avoid adverbs and passive voice, can I suggest:

While `oc` commands can perform the same tasks, you can use `svcat` to allow easier interaction with Service Catalog resources.

bfallonf commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,438 @@
+
+[[service-catalog-cli-find-broker]]
+==== Find brokers
+To view all the brokers installed on the cluster, run the following command:
```
Because #minimalism, maybe drop the "run the following command"?

bfallonf replied:
Happens a few more times in this PR.

bfallonf commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,438 @@
+[[service-catalog-cli-view-serviceplans]]
+==== View service plans
+To view the ClusterServicePlan resources available in the cluster, run the
+command:
```
Same as above.

bfallonf commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,438 @@
+
+[[service-catalog-cli-delete-servicebrokers]]
+=== Deleting service brokers
+. To remove broker services for the service catalog, delete the `ClusterServiceBroker` resource.
```
: at the end

bfallonf commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,438 @@
+  ansible-service-broker    https://asb.openshift-ansible-service-broker.svc:1338/ansible-service-broker                Ready
+----
+
+. View the `ClusterServiceClass` resources for the broker, to verify that the broker is removed.
```
Same

bfallonf commented:
@gaurav-nelson I made some nits :cow2:

gaurav-nelson commented:
Thanks @bfallonf 💵

jpeeler commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,435 @@
++
+[source, bash]
+----
+$ svcat unbind -n <project-name> <instance-name>
```
This part doesn't seem to flow well to me. Since 1 and 2 are really the same step, I think they should be combined.

Also, it should be made clear that that when bindings are removed this way, the user doesn't have to explicitly name each binding to remove. But then give an example of how that is done, such as with `svcat unbind -n szh-project --name mediawiki-postgresql-binding`.

gaurav-nelson commented:
Thanks @jpeeler 

I have updated the example and added a note with additional command to delete individual bindings.

![image](https://user-images.githubusercontent.com/23069445/41324031-8ea77006-6ef5-11e8-8b3a-cbcba6cb0607.png)

sunzhaohua2 commented on architecture/service_catalog/service_catalog_cli.adoc:
```diff
@@ -0,0 +1,446 @@
++
+[source, bash]
+----
+$ svcat get classes
```
The output should be as below, uuid have been removed. 
```
$ svcat get classes
  NAME   DESCRIPTION
+------+-------------+
```

gaurav-nelson replied:
Thanks @sunzhaohua2 I have updated this.

gaurav-nelson commented:
/cherrypick enterprise-3.10
